docker/service.go

```diff
@@ -344,21 +358,58 @@ func (s *Service) collectContainersAndDo(ctx context.Context, action func(*Conta
 }
 
 func (s *Service) eachContainer(ctx context.Context, containers []*Container, action func(*Container) error) error {
-
 	tasks := utils.InParallel{}
 	for _, container := range containers {
-		task := func(container *Container) func() error {
-			return func() error {
-				return action(container)
-			}
-		}(container)
+		task := taskFunc(container, action)
+		tasks.Add(task)
+	}
+	return tasks.Wait()
+}
+
+func (s *Service) getFilters() filters.Args {
+	filter := filters.NewArgs()
+	filter.Add("label", fmt.Sprintf("%s=%s", labels.PROJECT, s.project.Name))
+	filter.Add("label", fmt.Sprintf("%s=%s", labels.SERVICE, s.name))
+	return filter
+}
 
+func (s *Service) watchContainers(ctx context.Context, initialContainers []*Container, action func(*Container) error, dieAction func(*Container) error) error {
+	tasks := utils.InParallel{}
+	for _, container := range initialContainers {
+		task := taskFunc(container, action)
 		tasks.Add(task)
 	}
 
+	filter := s.getFilters()
+	cli := s.clientFactory.Create(s)
+
+	handler := dockerevents.NewHandler(dockerevents.ByAction)
+	handler.Handle("die", func(m eventtypes.Message) {
+		container, _ := ExistingContainer(ctx, cli, m.ID, s)
+		if container != nil {
+			tasks.Add(taskFunc(container, dieAction))
+		}
+	})
+	handler.Handle("create", func(m eventtypes.Message) {
+		container, _ := ExistingContainer(ctx, cli, m.ID, s)
+		if container != nil {
+			tasks.Add(taskFunc(container, action))
+		}
+	})
+
+	dockerevents.MonitorWithHandler(ctx, cli, types.EventsOptions{
+		Filters: filter,
+	}, handler)
+
 	return tasks.Wait()
 }
 
+func taskFunc(container *Container, action func(*Container) error) func() error {
+	return func() error {
+		return action(container)
```

Comment:
``action`` function may also need ``context.Context`` to be passed, which will be helpful for finding related containers to service/project.